packages/react/menu/src/Menu.tsx

```diff
@@ -55,6 +98,106 @@ const Menu: React.FC<MenuOwnProps> = (props) => {
+  const context = useMenuContext(SUB_NAME);
+  const [focusFirst, setFocusFirst] = React.useState(false);
+  const [trigger, setTrigger] = React.useState<MenuSubTriggerElement | null>(null);
+  const [content, setContent] = React.useState<MenuContentElement | null>(null);
+  const contentId = useId();
+  const handleOpenChange = useCallbackRef(onOpenChange);
+
+  // If a parent menu unmounts, we ensure its submenus reset their open state.
+  // This prevents the parent menu from reopening with open submenus.
+  React.useEffect(() => {
+    if (context.open === false) handleOpenChange(false);
+    return () => handleOpenChange(false);
+  }, [context.open, handleOpenChange]);
+
+  React.useEffect(() => {
+    if (focusFirst) content?.focus();
+  }, [content, focusFirst]);
+
+  return (
+    <PopperPrimitive.Root>
+      <MenuProvider
+        isSubmenu={true}
+        dir={context.dir}
+        open={open}
+        onOpenChange={handleOpenChange}
+        trigger={trigger}
+        onTriggerChange={setTrigger}
+        content={content}
+        contentId={contentId}
+        onContentChange={setContent}
+        onRootClose={context.onRootClose}
+        onEntryFocus={React.useCallback(
+          (event) => {
+            if (!focusFirst) event.preventDefault();
+            setFocusFirst(false);
+          },
+          [focusFirst]
+        )}
+        onKeyOpen={React.useCallback(() => {
+          setFocusFirst(true);
+          handleOpenChange(true);
+        }, [handleOpenChange])}
+      >
+        {children}
+      </MenuProvider>
+    </PopperPrimitive.Root>
+  );
+};
+
+MenuSub.displayName = SUB_NAME;
+
+/* -------------------------------------------------------------------------------------------------
+ * MenuSubTrigger
+ * -----------------------------------------------------------------------------------------------*/
+
+const SUB_TRIGGER_NAME = 'MenuSubTrigger';
+
+type MenuSubTriggerOwnProps = Polymorphic.OwnProps<typeof MenuItemImpl>;
+type MenuSubTriggerPrimitive = Polymorphic.ForwardRefComponent<
+  Polymorphic.IntrinsicElement<typeof MenuItemImpl>,
+  MenuSubTriggerOwnProps
+>;
+
+const MenuSubTrigger = React.forwardRef((props, forwardedRef) => {
+  const context = useMenuContext(SUB_TRIGGER_NAME);
+  return context.isSubmenu ? (
+    <MenuAnchor as={Slot}>
+      <MenuItemImpl
+        aria-haspopup="menu"
+        aria-expanded={context.open || undefined}
+        aria-controls={context.contentId}
+        data-state={getOpenState(context.open)}
+        {...props}
+        ref={composeRefs(forwardedRef, context.onTriggerChange)}
+        onMouseMove={composeEventHandlers(props.onMouseMove, () => {
+          if (!props.disabled) context.onOpenChange(true);
+        })}
+        onKeyDown={composeEventHandlers(props.onKeyDown, (event) => {
+          if (!props.disabled && SUB_OPEN_KEYS[context.dir].includes(event.key)) {
+            context.onKeyOpen();
+          }
+        })}
+      />
+    </MenuAnchor>
+  ) : null;
```

Comment:
Are we doing this in case someone was to add a sub trigger where it's not supposed to be?
I wonder if we need to bother guarding against that considering it's a private component for us to use in `DropdownMenu` and `ContextMenu`.

Reply:
We're doing it because we need to access properties on the context that only exist in `MenuSubContextValue`. So this helps TS narrow the union and an added bonus is that it prevents it from being rendered where it's not supposed to.

Reply:
ha ok cool